### PR TITLE
[SPARK-29722][SQL] Non reversed keywords should be able to be used in high order functions

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Support non-reversed keywords to be used in high order functions.

### Why are the changes needed?

the keywords are non-reversed.

### Does this PR introduce any user-facing change?

yes, all non-reversed keywords can be used in high order function correctly

### How was this patch tested?

add uts

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @ueshin @gatorsmile @cloud-fan @HyukjinKwon @dongjoon-hyun, thanks in advance.

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -718,8 +718,8 @@ primaryExpression
     | '(' query ')'                                                                            #subqueryExpression
     | qualifiedName '(' (setQuantifier? argument+=expression (',' argument+=expression)*)? ')'
        (OVER windowSpec)?                                                                      #functionCall
-    | IDENTIFIER '->' expression                                                               #lambda
-    | '(' IDENTIFIER (',' IDENTIFIER)+ ')' '->' expression                                     #lambda
+    | identifier '->' expression                                                               #lambda
+    | '(' identifier (',' identifier)+ ')' '->' expression                                     #lambda
```
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Ur, I see. Looks reasonable to me.

File: sql/core/src/test/resources/sql-tests/inputs/higher-order-functions.sql [hunk 2]
```diff
@@ -12,9 +12,14 @@ select transform(zs, z -> z) as v from nested;
...
 -- Transform an array
 select transform(ys, y -> y * y) as v from nested;
+-- use non reversed keywords
+select transform(ys, left -> left * left) as v from nested;
+
 
 -- Transform an array with index
 select transform(ys, (y, i) -> y + i) as v from nested;
+-- use non reversed keywords
+select transform(ys, (cost, i) -> cost + i) as v from nested;
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/higher-order-functions.sql]: Can you move the two tests above into the end of the file to make diff smaller in `sq.out`.

Comment (maropu) [reply sql/core/src/test/resources/sql-tests/inputs/higher-order-functions.sql]: Also, can you add some tests for error cases if ansi=true?

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/higher-order-functions.sql]: thanks for you suggestion. I just moved them to the end of this file, and add some cases to verify with ansi flag.

**Review by maropu — APPROVED** [review]
LGTM cc: @ueshin @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**